### PR TITLE
Support Aerospike

### PR DESCRIPTION
**Pull request by tlopatic**
Hey guys,

We've been using YCSB quite a bit at Aerospike lately. It would be nice, if we could get support for our database into the upstream YCSB code.

Thomas

### CODE REVIEW

File: aerospike/src/main/java/com/yahoo/ycsb/db/AerospikeClient.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+public class AerospikeClient extends com.yahoo.ycsb.DB {
+  private static final boolean DEBUG = false;
+
+  private static final String DEFAULT_HOST = "localhost";
+  private static final String DEFAULT_PORT = "3000";
+  private static final String DEFAULT_TIMEOUT = "10000";
+  private static final String DEFAULT_NAMESPACE = "ycsb";
+
+  private static final int RESULT_OK = 0;
+  private static final int RESULT_ERROR = 1;
```
Comment (cmccoy) [inline aerospike/src/main/java/com/yahoo/ycsb/db/AerospikeClient.java]: These should be unified eventually, but negative values seem more prevalent for errors in most YCSB clients - consider using -1.

File: aerospike/src/main/java/com/yahoo/ycsb/db/AerospikeClient.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        if (DEBUG) {
+          System.err.println("Record key " + key + " not found (delete)");
+        }
+
+        return RESULT_ERROR;
+      }
+
+      return RESULT_OK;
+    } catch (AerospikeException e) {
+      System.err.println("Error while deleting key " + key + ": " + e);
```
Comment (cmccoy) [inline aerospike/src/main/java/com/yahoo/ycsb/db/AerospikeClient.java]: Wrap this in an `if(DEBUG)`?

File: aerospike/src/main/java/com/yahoo/ycsb/db/AerospikeClient.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    Key keyObj = new Key(namespace, table, key);
+
+    while (true) {
+      try {
+        client.put(writePolicy, keyObj, bins);
+        writeOverloadTries = WRITE_OVERLOAD_TRIES;
+        return RESULT_OK;
+      } catch (AerospikeException e) {
+        if (e.getResultCode() != ResultCode.DEVICE_OVERLOAD) {
+          System.err.println("Error while writing key " + key + ": " + e);
```
Comment (cmccoy) [inline aerospike/src/main/java/com/yahoo/ycsb/db/AerospikeClient.java]: `if(DEBUG)?`

**Comment by cmccoy** [comment]
Nice! This will be great to have. Left a few comments.

**Comment by tlopatic** [comment]
This is exciting!

Thanks for the constructive feedback.

A bit on the DEBUG flag: Yes, its use isn't entirely consistent right now. Generally, the idea is to use it to distinguish between expected/recoverable and unexpected errors.

Accordingly, the "Record key ... not found" error in read() should not be wrapped in DEBUG. We should never run into a situation where we read a record that doesn't exist. However, there seems to be a race condition in workload D that causes exactly that.

The client threads generate keys for the records to be inserted and there is a small time window between the thread generating the key and it actually inserting the record. However, as soon as the key has been generated, that key is considered to be available for other threads to read. (Because transactioninsertkeysequence.lastInt() now returns that key.) That results in the other threads potentially trying to read records that don't yet exist.

The right fix is, obviously, to take care of the race condition. I just didn't have time, yet, to think about this a little more in depth.

So, for now, I am mimicking the behavior of other drivers, which don't output an error message and just return an error code.

Or were you referring to something else regarding the flag?

Oh, and let me open a ticket regarding the race condition, so that it gets tracked.

**Comment by tlopatic** [comment]
Oh, and I changed the error result to -1.

**Comment by busbey** [comment]
whoever merges this, please not in the merge commit that it also closes #113

File: aerospike/src/main/java/com/yahoo/ycsb/db/AerospikeClient.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+  @Override
+  public int scan(String table, String start, int count, Set<String> fields,
+      Vector<HashMap<String, ByteIterator>> result) {
+    System.err.println("Scan not implemented");
+    return RESULT_ERROR;
+  }
+
+  private int write(String table, String key, WritePolicy writePolicy,
+      HashMap<String, ByteIterator> values) {
+    if (writeOverloadTries == 0) {
```
Comment (cmccoy) [inline aerospike/src/main/java/com/yahoo/ycsb/db/AerospikeClient.java]: A little description of the retry policy here would be very helpful :).

Is the intent here that once a write fails `WRITE_OVERLOAD_TRIES` consecutive times with error code `DEVICE_OVERLOAD`, no further writes will be issued by the client?

**Comment by cmccoy** [comment]
@tlopatic  - looks great. Left one other comment, if you don't mind.

**Comment by tlopatic** [comment]
Sorry for the delay. I was traveling.

Yes, your interpretation of the retry policy is correct. And, thinking about it again, it's a little... idiosyncratic. It did fit the specific way in which I was using YCSB, but it was probably a little counter-intuitive for everybody else.

So, instead of documenting its behavior, let's correct the behavior, so that it makes more sense.

Now each write is tried 3 times, followed by delays of 5, 10, and 20 ms, respectively. That's probably a little less surprising than the previous behavior. It'll allow people to see what fraction of their writes actually failed because of server overload. (Instead of failing every subsequent write, once an overload situation is encountered.)

Sound good?

**Comment by cmccoy** [comment]
Great!

I do think it's a little counterintuitive that *any* retries happens within YCSB, but the Cassandra Thrift client also contains retry logic.  Opened #334 to discuss, merging this.

Thanks again, @tlopatic!

**Pull request merged**